omnipaxos_storage/src/persistent_storage.rs

```diff
@@ -0,0 +1,384 @@
+    /// the path to the directory containing a commitlog
+    log_path: String,
+    /// local RocksDB key-value store
+    rocksdb: DB,
+    /// A placeholder for the T: Entry
+    t: PhantomData<T>,
+    /// A placeholder for the S: Snapshot<T>
+    s: PhantomData<S>,
+}
+
+impl<T: Entry, S: Snapshot<T>> PersistentStorage<T, S> {
+    pub fn with(storage_config: PersistentStorageConfig) -> Self {
+        // Paths to commitlog and rocksDB store
+        let log_path: String = storage_config.commitlog_path.unwrap();
+        let rocks_path = storage_config.rocksdb_path.unwrap();
+
+        // Check if storage already exists on the paths
+        std::fs::metadata(&log_path)
+            .expect_err(&format!("commitlog already exists in {}", log_path));
+        std::fs::metadata(&rocks_path)
+            .expect_err(&format!("rocksDB store already exists in {}", rocks_path));
+
+        // Initialize Commitlog and rocksDB
+        let commitlog =
+            CommitLog::new(storage_config.commitlog_options).expect("Failed to create Commitlog");
+        let rocksdb = DB::open(&storage_config.rocksdb_options, &rocks_path)
+            .expect("Failed to create rocksDB store");
+
+        Self {
+            commitlog,
+            log_path,
+            rocksdb,
+            t: PhantomData::default(),
+            s: PhantomData::default(),
+        }
+    }
+}
+
+impl<T, S> Storage<T, S> for PersistentStorage<T, S>
+where
+    T: Entry + Serialize + for<'a> Deserialize<'a>,
+    S: Snapshot<T> + Serialize + for<'a> Deserialize<'a>,
+{
+    fn append_entry(&mut self, entry: T) -> u64 {
+        let entry_bytes = bincode::serialize(&entry).expect("Failed to serialize");
+        self.commitlog
+            .append_msg(entry_bytes)
+            .expect("Failed to append log entry")
+            + 1
+    }
+
+    fn append_entries(&mut self, entries: Vec<T>) -> u64 {
+        let serialized = entries
+            .into_iter()
+            .map(|entry| bincode::serialize(&entry).expect("Failed to serialize"));
+        let offset = self
+            .commitlog
+            .append(&mut MessageBuf::from_iter(serialized))
+            .expect("Falied to append entries");
+        offset.first() + offset.len() as u64
+    }
+
+    fn append_on_prefix(&mut self, from_idx: u64, entries: Vec<T>) -> u64 {
+        self.commitlog
+            .truncate(from_idx)
+            .expect("Failed to truncate log");
+        self.append_entries(entries)
+    }
+
+    fn get_entries(&self, from: u64, to: u64) -> Vec<T> {
+        let buffer = self
+            .commitlog
+            .read(from, ReadLimit::default())
+            .expect("Failed to read from log");
+        let mut entries = Vec::<T>::with_capacity((to - from) as usize);
+        let mut iter = buffer.iter();
+        for _ in from..to {
+            let msg = iter.next().unwrap();
+            entries.push(bincode::deserialize(msg.payload()).expect("Failed to deserialize"));
```

Comment:
get_entries should either return either all entries between `from` and `to` if they exist in the log. If not, it should return an empty vec. Thus, this needs to be changed to something like this:
```suggestion
            match iter.next() {
                   Some(msg) => {
                        entries.push(bincode::deserialize(msg.payload()).expect("Failed to deserialize"));
                   },
                   None => {
                        return vec![];   // early return here
                   }     
            }
```

I also added a comment in the description for this function in the `Storage` trait to make it clear for others too.